<a href="https://colab.research.google.com/github/arcila-heisman/ml_introductory_labs/blob/master/Lab_2/Lab_2_Parte2_No_Parametrico.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Laboratorio 2 - Parte 2

### Modelos no paramétricos

### 2019-I

#### Profesor: Julián D. Arias Londoño
#### julian.ariasl@udea.edu.co

Modificado por Heisman Arcila. (Versión original en: https://github.com/jdariasl/ML_IntroductoryCourse/tree/master/Labs)

## Guía del laboratorio

En esta archivo va a encontrar tanto celdas de código cómo celdas de texto con las instrucciones para desarrollar el laboratorio.

Lea atentamente las instrucciones entregadas en las celdas de texto correspondientes y proceda con la solución de las preguntas planteadas.

Nota: no olvide ir ejecutando las celdas de código de arriba hacia abajo para que no tenga errores de importación de librerías o por falta de definición de variables.

#### Primer integrante:

Nombre: Danny Francisco Hernandez Godoy

CC: 1105784633

#### Segundo integrante:

Nombre: Daniel Martinez

CC: 1036645337

In [0]:
from __future__ import division
%matplotlib inline
import numpy as np
import scipy as sc
import matplotlib.pyplot as plt


#Evitar algunas advertencias
import warnings
warnings.filterwarnings("always")

## Ejercicio 1: Contextualización del problema


Usaremos el dataset iris para el problema de clasificación, desde la librería sklearn. En el UCI Machine Learning Repository se encuentra más información en el siguiente [link](https://archive.ics.uci.edu/ml/datasets/iris) .

In [0]:
#cargamos la bd iris desde el dataset de sklearn
from sklearn import datasets

iris = datasets.load_iris()
X, Y = iris.data, iris.target

Responda las siguientes preguntas: 

1.1 ¿Cu&aacute;ntas clases tiene el problema?: Tiene 3 clases


In [0]:
print(np.unique(Y))

[0 1 2]


1.2 ¿Cu&aacute;ntas caracter&iacute;sticas tiene el problema?: Tiene 4 caracteristicas

In [0]:
print(X.shape)

(150, 4)


1.3 ¿Cu&aacute;ntas muestras tiene el problema?: Tiene 150 muestras

In [0]:
print(X.shape)

(150, 4)


1.4 ¿Cu&aacute;ntas muestras por cada clase hay en la base de datos?: 50 muestras por cada uno

In [0]:
print((Y))

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


## Ejercicio 2: Completar código KNN

Analice los métodos de la teoría vista para los modelos de *K-vecinos más cercanos (KNN)* y complete el código del método KNN.

<b>Nota</b>: Para el cáculo de la distancia entre vectores tienen dos opciones, usar la función la distancia euclidiana `scipy.spatial.distance.euclidean`([Ejemplo](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.euclidean.html)) o usar la función `numpy.linalg.norm`([Ejemplo](https://docs.scipy.org/doc/numpy/reference/generated/numpy.linalg.norm.html)). Revisen la documentación para comprender su uso. También serán de utilidad las funciones `numpy.sort` y `numpy.argsort`.


In [0]:
from scipy.stats import mode
import operator

#Para calcular el error en los problemas de clasificación
def ErrorClas(Y_lest, Y):
    error = 1 - np.sum(Y_lest == Y)/len(Y)
    
    return error


def distance(X_val, X_train):
    
    MdotXY = np.dot(X_val, X_train.T)
    sumxv = np.sum(X_val * X_val, axis=1).reshape([X_val.shape[0],1])
    sumxt = np.sum(X_train * X_train, axis=1).reshape([1,X_train.shape[0]])
    
    D = np.dot(sumxv, np.ones((1,X_train.shape[0]))) + np.dot(np.ones((X_val.shape[0],1)), sumxt) - 2* MdotXY
    D = np.sqrt(D)
    
    return D

def neighborhood (k, D_argsort, Y_train):
    N = len(D_argsort)    
    kn = D_argsort[ : , 0: k]
    kn = kn.reshape(len(D_argsort),k)
    Y_val = np.zeros((N,k))

  #  Y_val =np.array( [[Y_train[kn[y,x]] for x in range(k)] for y in range(N)])
    for x in range (N):
        for y in range (k):
               Y_val[x][y] = Y_train[kn[x,y]]
    
    return Y_val

def KNN(X_train, Y_train, X_val, k, tipo):
    
    '''
    X_train es la matriz con las muestras de entrenamiento
    Y_train es un vector con los valores de salida pra cada una de las muestras de entrenamiento
    X_val es la matriz con las muestras de validación
    tipo es una bandera que indica si el problema es de regresión o de clasificación.
    '''
    k = k   #Parámetro k que equivale al número de vecinos a tener en cuenta para resolver el problema de predicción de la variable de salida
    
    N = np.shape(Y_train)[0]
    Y_val = np.zeros(N).reshape(N,1)
    Y_train = Y_train.reshape(N,1)
    
    D = distance(X_val, X_train)
    D_argsort = np.argsort(D, axis = 1)
    
    LXv = len(X_val)
    
    Y_argsort = neighborhood(k, D_argsort, Y_train)

    Y_val = np.zeros(LXv).T
    

    if(tipo == 'r'): #Regresion
        #regresion
        for x in range (Nt):
            average = np.average(Y_argsort[x])
            Y_val[x] = average
            
    elif(tipo == 'c'): #clasificacion
        #clasificacion
        
        for x in range (LXv):
            mode = stats.mode(Y_argsort[x])           
            Y_val[x] = mode[0].astype(float)
            
    Y_val.reshape(LXv,1)
    
    return Y_val #Debe retornar un vector que contenga las predicciones para cada una de las muestras en X_val, en el mismo orden.  

## Ejercicio 3

Una vez haya completado el código del método de KNN, ejecute varias veces el proceso de entrenamiento y evaluación cambiando el parametro $k$, el cual es el número de vecinos, y complete la siguiente tabla con los valores del error de clasificación obtenidos:

In [0]:
from numpy import random
from sklearn.model_selection import train_test_split
import math
from scipy import stats

N = np.size(X,0)
#KK=2
def invento(KK):
  #Validamos el modelo
  random.seed(1)
  Error = np.zeros(4)
  for j in range(4):

      Xtrain, Xtest, Ytrain, Ytest = train_test_split(X, Y, test_size=0.25)

      #Normalizamos los datos
      media = np.mean(Xtrain)
      desvia = np.std(Xtrain)
      Xtrain = sc.stats.stats.zscore(Xtrain)
      Xtest = (Xtest - np.matlib.repmat(media, Xtest.shape[0], 1))/np.matlib.repmat(desvia, Xtest.shape[0], 1)

      Yest = KNN(Xtrain, Ytrain, Xtest, KK, 'c')#Complete con el llamado apropiado de la función que Uds diseñaron

      #Evaluamos las predicciones del modelo con los datos de test
      Error[j] = ErrorClas(Yest,Ytest)
  print('\nError durante la prueba = ' + str(np.mean(Error)) + '+-' + str(np.std(Error)))
  ee=np.mean(Error)
  ed=np.std(Error)
  return ee,ed, Xtrain, Xtest

Responda:

3.1 ¿Qué metodología de validación se usa en el experimento? Validación bootstraping ya que se usa un 74,6% en el entrenamiento y un 25,4% en la validación


In [0]:
TT=invento(2)
print(TT[2])
print(TT[3])


Error durante la prueba = 0.26973684210526316+-0.08802031684381349
[[-0.16691592  2.92889951 -1.29067203 -1.04419434]
 [-1.01666972  0.49407887 -1.348058   -1.30938656]
 [ 1.41119827  0.27273153  0.54567912  0.28176673]
 [-0.89527632  1.60081552 -1.29067203 -1.17679045]
 [ 0.31865767 -0.16996313  0.66045107  0.81215115]
 [ 0.31865767 -0.61265779  0.54567912  0.01657451]
 [-1.50224331  0.27273153 -1.348058   -1.30938656]
 [-0.04552252 -0.61265779  0.77522302  1.6077278 ]
 [ 0.56144447 -1.27669979  0.66045107  0.41436283]
 [ 0.19726427  0.7154262   0.43090717  0.54695894]
 [ 1.04701807 -0.16996313  0.83260899  1.47513169]
 [ 0.19726427 -0.83400512  0.77522302  0.54695894]
 [-1.25945652 -0.16996313 -1.348058   -1.17679045]
 [ 0.19726427 -1.94074178  0.14397731 -0.2486177 ]
 [-1.13806312 -1.49804712 -0.25772451 -0.2486177 ]
 [-1.62363671 -1.71939445 -1.40544397 -1.17679045]
 [ 0.68283787  0.0513842   1.00476691  0.81215115]
 [-0.53109612 -0.16996313  0.43090717  0.41436283]
 [ 0.80423127 

3.2 ¿Cuántas muestras se usan para el entrenamiento? Se usan 112 muestras para entrenamiento


In [0]:
print(TT[2].shape)

(112, 4)


3.3 ¿Cuántas muestras se usan para la validación? Se usan 38 para validación

In [0]:
print(TT[3].shape)

(38, 4)


#### Tabla de resultados

In [0]:
import pandas as pd
#import qgrid
randn = np.random.randn
df_types = pd.DataFrame({
    'Numero de vecinos' : pd.Series(['1', '2', '3', '4', '5', '6', '7'])})
df_types["Error_Prueba"] = ""
df_types["Desviación estándar del error"] = ""
df_types.set_index(['Numero de vecinos'], inplace=True)

for i in df_types.index:
    print(i[0])
    EE=invento(int(i[0]))
    df_types["Error_Prueba"][i] = EE[0]
    df_types["Desviación estándar del error"][i] = EE[1]

#''' acá llene la tabla, tenga en cuenta que 
#df_types["Error_Prueba"][1] = "0.2697"
#df_types["Desviación estándar del error"][1] = "0.088" #para evaluear su código'''


#df_types.sort_index(inplace=True)
#qgrid_widget = qgrid.show_grid(df_types, show_toolbar=False)
#qgrid_widget

1

Error durante la prueba = 0.20394736842105263+-0.11376063464335907
2

Error durante la prueba = 0.26973684210526316+-0.08802031684381349
3

Error durante la prueba = 0.26973684210526316+-0.08603090020146066
4

Error durante la prueba = 0.28289473684210525+-0.07529949435697103
5

Error durante la prueba = 0.26973684210526316+-0.08802031684381349
6

Error durante la prueba = 0.2763157894736842+-0.0842516347030638
7

Error durante la prueba = 0.2763157894736842+-0.0842516347030638


##No funciona qgrid

In [0]:
df_types

Error_Prueba Desviación estándar del error
Numero de vecinos                                           
1                     0.203947                      0.113761
2                     0.269737                     0.0880203
3                     0.269737                     0.0860309
4                     0.282895                     0.0752995
5                     0.269737                     0.0880203
6                     0.276316                     0.0842516
7                     0.276316                     0.0842516

Ejecute la siguiente instrucción para dejar guardados en el notebook los resultados de las pruebas.

In [0]:
qgrid_widget.get_changed_df()

NameError: ignored

## Ejercicio 4

A continuación debe completar el código del método parzen para resolver problemas de clasificación con el modelo ventana de Parzen.

In [0]:
#Calcula el kernel gaussiano de x
def kernel_gaussiano(x):
    return np.exp((-0.5)*x**2)

def ParzenWindow(x,Data,h):
    h = h
    Ns = Data.shape[0]
    suma = 0
    for k in range(Ns):
        u = sc.spatial.distance.euclidean(x,Data[k,:])
        suma += kernel_gaussiano(u/h)
    return suma

def ParzenClass(X_train, Y_train, X_val, ancho_h):
    '''Complete el código para predecir las clases de la muestra x_val'''
    Lx = X_val.shape[0]
    Yest = np.zeros(Lx)
    h = ancho_h
    #print(h, "Aqui entro")
    #clases
    class0 = X_train[Y_train == 0]
    class1 = X_train[Y_train == 1]
    class2 = X_train[Y_train == 2]
    
    z0= np.zeros(Lx)
    z1= np.zeros(Lx)   
    z2= np.zeros(Lx)
    
    for i in range(Lx):        
        z0[i] = ParzenWindow(X_val[i],class0, h)
        z1[i] = ParzenWindow(X_val[i],class1, h)
        z2[i] = ParzenWindow(X_val[i],class2, h)        
        z = np.array([z0[i], z1[i], z2[i]])
        classM = np.argmax(z)
        Yest[i] = classM    
    return Yest#Debe retornar un vector que contenga las predicciones para cada una de las muestras en X_val, en el mismo orden.  

## Ejercicio 5

Ahora debe resolver el mismo problema de clasificación con el método ventana de Parzen. Complete el código ejecute las pruebas para diferentes valores de $h$ y llene la siguiente tabla:

In [0]:
from numpy import random
from sklearn.model_selection import StratifiedKFold
import math

def invento2(ancho):
  #Validamos el modelo
  random.seed(1)
  Error = np.zeros(4)
  skf = StratifiedKFold(n_splits=4)
  j = 0
  for train, test in skf.split(X, Y):
      Xtrain = X[train,:]
      Ytrain = Y[train]
      Xtest = X[test,:]
      Ytest = Y[test]
      #Normalizamos los datos
      media = np.mean(Xtrain)
      desvia = np.std(Xtrain)
      Xtrain = sc.stats.stats.zscore(Xtrain)
      Xtest = (Xtest - np.matlib.repmat(media, Xtest.shape[0], 1))/np.matlib.repmat(desvia, Xtest.shape[0], 1)

      Yest = ParzenClass(Xtrain, Ytrain, Xtest, ancho)#Complete con el llamado apropiado de la función que Uds diseñaron

      #Evaluamos las predicciones del modelo con los datos de test
      Error[j] = ErrorClas(Yest,Ytest)
      j += 1
  print('\nError durante la prueba = ' + str(np.mean(Error)) + '+-' + str(np.std(Error)))
  le=np.mean(Error)
  lx=np.std(Error)
  return le,lx, Xtrain, Xtest, Ytest
  

Responda:
    
5.1 ¿Qué metodología de validación se usó en la simulación?: Validacion cruzada estratificada, se usa K-1 para entrenar el sistema y lo otro para validar y se repite el proceso para los diferentes K


5.2 ¿Cuántas muestras se usan para el entrenamiento?: Se usan 114 muestras para el entrenamiento


5.3 ¿Cuántas muestras se usan para la validación?: Se usan 36 muestras para la validación

In [0]:
nota=invento2(0.1)
print(nota[2].shape, "Entrenamiento")
print(nota[3].shape, "Validacion")


Error durante la prueba = 0.2216880341880342+-0.04332854434165311
(114, 4) Entrenamiento
(36, 4) Validacion


#### Tabla de resultados

In [0]:
import pandas as pd
#import qgrid
randn = np.random.randn
df_types2 = pd.DataFrame({
    'Tamano de ventana' : pd.Series(['0.05', '0.1', '0.5', '1', '2', '5', '10'])})
df_types2["Error_Prueba"] = ""
df_types2["Desviación estándar del error"] = ""
df_types2.set_index(['Tamano de ventana'], inplace=True)
''' acá llene la tabla, tenga en cuenta que
df_types2["Error_Prueba"][1] = "0.222"
df_types2["Desviación estándar del error"][1] = "0.043" para evaluar su código'''
for i in df_types2.index:
    #print(float(i[0]))
    EX=invento2(float(i))
    df_types2["Error_Prueba"][i] = EX[0]
    df_types2["Desviación estándar del error"][i] = EX[1]

#df_types.sort_index(inplace=True)
#qgrid_widget = qgrid.show_grid(df_types2, show_toolbar=False)
#qgrid_widget


Error durante la prueba = 0.2216880341880342+-0.04332854434165311

Error durante la prueba = 0.2216880341880342+-0.04332854434165311

Error durante la prueba = 0.281517094017094+-0.04021339704891218

Error durante la prueba = 0.280982905982906+-0.03145818774430656

Error durante la prueba = 0.24198717948717952+-0.058466142137675314

Error durante la prueba = 0.2216880341880342+-0.04696907336459245

Error durante la prueba = 0.2216880341880342+-0.04696907336459245


##No funciona Qgrid

In [0]:
df_types

Error_Prueba Desviación estándar del error
Numero de vecinos                                           
1                     0.203947                      0.113761
2                     0.269737                     0.0880203
3                     0.269737                     0.0860309
4                     0.282895                     0.0752995
5                     0.269737                     0.0880203
6                     0.276316                     0.0842516
7                     0.276316                     0.0842516

Ejecute la siguiente instrucción para dejar guardados en el notebook los resultados de las pruebas.

In [0]:
qgrid_widget.get_changed_df()

Error_Prueba Desviación estándar del error
Tamaño de ventana                                           
0.05                                                        
0.1                      0.222                         0.043
0.5                                                         
1                                                           
2                                                           
5                                                           
10